src/server/debug.rs

```diff
@@ -977,6 +1036,100 @@ impl<ER: RaftEngine> Debugger for DebuggerImpl<ER> {
         self.reset_to_version_manager.start(version.into());
     }
 
+    fn region_flashback_to_version(
+        &self,
+        region_id: u64,
+        version: u64,
+        start_ts: TimeStamp,
+        commit_ts: TimeStamp,
+    ) -> impl Future<Output = Result<()>> + Send {
+        let store_id = self.get_store_ident().unwrap().get_store_id();
+        let r = self.region_info(region_id).unwrap();
+        let region = r
+            .region_local_state
+            .as_ref()
+            .map(|s| s.get_region().clone())
+            .unwrap();
+
+        let mut ctx = Context::default();
+        ctx.set_region_id(region.get_id());
+        ctx.set_region_epoch(region.get_region_epoch().to_owned());
+        let peer = find_peer(&region, store_id).unwrap();
+        ctx.set_peer(peer.clone());
+
+        // Flashback will encode the key, so we need to use raw key.
+        let start_key = Key::from_encoded_slice(region.get_start_key())
+            .to_raw()
+            .unwrap_or_default();
+        let end_key = Key::from_encoded_slice(region.get_end_key())
+            .to_raw()
+            .unwrap_or_default();
+
+        let storage = self.storage.clone().unwrap();
+        let is_in_flashback = region.get_is_in_flashback();
+        async move {
+            let timeout = Duration::from_secs(FLASHBACK_TIMEOUT);
+            let start_time = Instant::now();
+
+            // Loop for single region.
+            loop {
+                let elapsed_time = start_time.elapsed();
+                if elapsed_time >= timeout {
+                    error!("Execute flashback timeout. Exiting the loop.");
+                    return Err(Error::Other(
+                        "Flashback timeout reached. Exiting the loop.".into(),
+                    ));
+                }
+
+                let storage_clone = storage.clone();
+                // Means now is prepare flashback.
+                if commit_ts.is_zero() {
+                    if is_in_flashback {
+                        return Ok(());
+                    }
+                    let mut req = kvrpcpb::PrepareFlashbackToVersionRequest::new();
+                    req.set_version(version);
+                    req.set_start_key(start_key.clone());
+                    req.set_end_key(end_key.clone());
+                    req.set_context(ctx.clone());
+                    req.set_start_ts(start_ts.into_inner());
+
+                    let resp = future_prepare_flashback_to_version(storage_clone, req)
+                        .await
+                        .unwrap();
+                    if !resp.get_error().is_empty() || resp.has_region_error() {
```

Comment:
When encountering the `EpochNotMatch` error, we have to re-check the region ID and its key range as well.

Reply:
done.